packages/core/src/read/impl.js

```diff
@@ -90,16 +90,33 @@ export async function performRead(context, readParams) {
   } = context.subsystems.read.context
 
   const kernel = context
+  const initialValue = kernel.query()
 
   const { uri, opts } = readParams
   const reader = registry.get(uri) || registry.get(fallback)
 
   if (reader != null) {
-    const readResponse = await time(`TIME-reader-(${uri})`, reader)(
-      uri,
-      opts,
-      R.pick(['config', 'subsystems', 'subsystemSequence'], context)
-    )
+    let enhanceContext = {
+      config: kernel.config,
+      subsystems: kernel.subsystems,
+      subsystemSequence: kernel.subsystemSequence,
+      elementZipper: kernel.elementZipper,
+    }
+
+    const [readResponse, contextBasedEnhancements] = await time(
+      `TIME-reader-plus-enhancement-(${uri})`,
+      Promise.all
+    )([
+      time(`TIME-reader-(${uri})`, reader)(
+        uri,
+        opts,
+        R.pick(['config', 'subsystems', 'subsystemSequence'], context)
+      ),
+      time(
+        `TIME-enhancement-extract-context-based-(${uri})`,
+        enhancement.extractContextBased
+      )(initialValue, enhanceContext),
```

Comment:
There's no initial value here, the "independent" enhancers take a single argument -- the context.

Reply:
The initial value is needed for the extraction of the kind. `loc` is this `initialValue`: https://github.com/netceteragroup/girders-elements/pull/91/files#diff-16e13b47bd24845124d598ca03fe174bR15.

Reply:
But you don't know what the kind of the received document will be *before* it arrives. Relying on our`__loading` convention for reads isn't reliable enough.

Perhaps a change in the API is in order:
- add the uri and opts to the enhanceContext 
- register enhancers on URI patterns instead of element kinds

@andon ?

Reply:
Ok, change of plan: 

The enhancer API si now:

```
enhance.register(parallelEnhance)
enhance.register(kind, enhance)
```

The signature of `parallelEnhance` is `parallelEnhance :: async context => updateFn || patternSet`
The signature of `enhance` is `enhance :: async (rootEl, context) => updateFn || patternSet`

A read runs *all* registered parallel enhancers together with the read. The "normal" ones are run as before. The updates are *applied* as now, all at once, after the "normal" enhancers finish.

Reasoning: it's better to simplify the API (i.e. not to rely on a optional part of the __read / __loading kind) rather than to fulfill the requirement. We could have allowed registrations on uri patterns, but then enhancers would really not add anything that a custom read can't achieve.

Reply:
@bevkoski yes, if used unwisely, this could lead to performance degradations because all parallel enhancers run indiscriminately. However, if you examine the use cases we have, for such enhancers, they are indiscriminate already (find all "action-buttons" and inject X; find all "articles" and inject Y), so somehow the new definition fits. If you combine it with a *run at most once* optimization (which we do) and that they run *in parallel with the main read*, which means that `totalTimeSpent = max(readTime, max(ehancerTime))`-- it will be ok.

Additinally, if you really like, you could short-cut the execution of the enhancer body if you are able to realize it is unnecessary based on the request (you should put the `uri` and `opts` in the context as well). 


On the conceptual side, I don't like complicated APIs (previously it would  have worked only if the hint of the expected kind [`__read, document, article`] would be present [it doesn't have to] *and* if the returned document kind matches the hint [it also doesn't necessarily have to).